src/wait_group.cr

```diff
@@ -0,0 +1,99 @@
+require "fiber"
+require "crystal/spin_lock"
+require "crystal/pointer_linked_list"
+
+# Suspend execution until a collection of fibers are finished.
+#
+# The wait group is a declarative counter of how many concurrent fibers have
+# been started. Each such fiber is expected to call `#done` to report that hey
+# are finished doing their work. Whenever the counter reaches zero the waiters
+# will be resumed.
+#
+# This is a simpler and more efficient alternative to using a `Channel(Nil)`
+# then looping a number of times until we received N messages to resume
+# execution.
+#
+# Basic example:
+#
+# ```
+# require "wait_group"
+# wg = WaitGroup.new(5)
+#
+# 5.times do
+#   spawn do
+#     do_something
+#   ensure
+#     wg.done # the fiber has finished
+#   end
+# end
+#
+# # suspend the current fiber until the 5 fibers are done
+# wg.wait
+# ```
+class WaitGroup
+  private struct Waiting
+    include Crystal::PointerLinkedList::Node
+
+    def initialize(@fiber : Fiber)
+    end
+
+    def enqueue : Nil
+      @fiber.enqueue
+    end
+  end
+
+  def initialize(n : Int32 = 0)
+    @waiting = Crystal::PointerLinkedList(Waiting).new
+    @lock = Crystal::SpinLock.new
+    @counter = Atomic(Int32).new(n)
+  end
+
+  # Increments the counter by how many fibers we want to wait for.
+  #
+  # This can also be used to decrement the counter, in which case the behavior
+  # is identical to `#done`.
+  #
+  # Can be called at any time, allowing concurrent fibers to add more fibers to
+  # wait for, but they must always do so before calling `#done` that would
+  # decrement the counter, to make sure that the counter may never inadvertently
+  # reach zero before all fibers are done.
+  def add(n : Int32 = 1) : Nil
+    new_value = @counter.add(n) + n
+    raise "Negative WaitGroup counter" if new_value < 0
+    return unless new_value == 0
+
+    @lock.sync do
+      @waiting.consume_each do |node|
+        node.value.enqueue
+      end
+    end
+  end
+
+  # Decrements the counter by one. Must be called by concurrent fibers once they
+  # have finished processing. When the counter reaches zero, all waiting fibers
+  # will be resumed.
+  def done : Nil
+    add(-1)
+  end
+
+  # Suspends the current fiber until the counter reaches zero, at which point
+  # the fiber will be resumed.
+  #
+  # Can be called from different fibers.
+  def wait : Nil
```

Comment:
I like this donation very much! It will be very useful in many cases.

One proposal that probably can be done later on as a separate improvement - is to make the `wait` method compatible with `select` to support the following snippet: 

```crystal
select
  when wg.wait
    puts "All fibers done"
  when timeout(X.seconds)
    puts "Some fiber stuck"
end
```

Reply:
@bararchy We're missing a generic mechanism for timeouts... but we could abstract how it's implemented for `select` so that could be doable.

That doesn't mean we can't _also_ integrate with `select`: we could wait on channel(s) + waitgroup(s) + timeout. Now, I'm not sure how to do that.

Reply:
I have a [commit to support WaitGroup in `select` expressions](https://github.com/crystal-lang/crystal/commit/dcfd2ad6bd7dd05b74a4db9cd0b6febbd035d339).

The integration wasn't complex after I understood how SelectAction and SelectContext are working, but the current implementation is very isolated to Channel (on purpose). Maybe the integration is not a good idea, but if proves to be a good idea, we might want to extract the `select` logic from `Channel` to the `Crystal` namespace.

I'll open a pull request after this one is merged, so we can have a proper discussion.

Reply:
@ysbaddaden now that it's in and merged, are you planning to make the followup PR? :eye: